File: aoa_server/game/Room.cpp
#include "Room.h"
#include "../communication/MessageSerializer.h"
#include "../core/Logger.h"
#include "GameDifficulty.h"
#include "GameStatus.h"
#include "Game.h"

Room::Room() {
    this->init();
}

Room::Room(int id) {
    this->init();
    this->id = id;
}

Room::Room(GameType type, GameDifficulty difficulty, BoardDimension dimension) {
    this->init();
    this->type = type;
    this->difficulty = difficulty;
    this->boardDimension = dimension;
}

void Room::init() {
    this->turn = 0;
    this->winnerID = -1;
    this->activePlayerIndex = 0;
    this->status = GameStatus::CONNECTING;
}

void Room::restart() {
    this->turn = 0;
    this->winnerID = -1;
    this->progress = std::queue<int>();
    this->updateActivePlayer();
}

int Room::getID() {
    return this->id;
}

void Room::setID(int id) {
    this->id = id;
}

int Room::getPlayerCount() {
    return (int) this->players.size();
}

BoardDimension Room::getBoardDimension() const {
    return this->boardDimension;
}

void Room::setBoardDimension(BoardDimension boardDimension) {
    this->boardDimension = boardDimension;
}

int Room::getActivePlayerID() {
    return this->playerOrder[this->activePlayerIndex];
}

void Room::updateActivePlayer() {
    this->activePlayerIndex = this->getPlayerCount() > 0 ? this->turn % this->getPlayerCount() : 0;
}

GameDifficulty Room::getDifficulty() const {
    return this->difficulty;
}

void Room::setDifficulty(GameDifficulty difficulty) {
    this->difficulty = difficulty;
}

PlayerMap &Room::getPlayers() {
    return this->players;
}

PlayerMap Room::copyPlayers() {
    return this->players;
}

void Room::registerPlayer(Player *p) {
    this->players[p->getID()] = p;
    this->playerOrder.push_back(p->getID());

    if(this->isRoomFull()) {
        if(this->isEverybodyOnline()) {
            this->changeStatus(GameStatus::READY);
        } else {
            this->changeStatus(GameStatus::CONNECTING);
        }
    } else {
        //Logger::info("++++++++++++++ NOT FULL");
    }
}

GameType Room::getGameType() {
    return this->type;
}

void Room::setGameType(GameType type) {
    this->type = type;
}

/**
 * Are all players assigned?
 * @return
 */
bool Room::isRoomFull() {
    return this->getPlayerCount() >= (int) this->getGameType();
}

/**
 * Is everybody in a room online?
 * @return
 */
bool Room::isEverybodyOnline() {
    return this->countOnlinePlayers() == this->getPlayerCount();
}

/**
 *
 * @return
 */
bool Room::isJoinable() {
    return !this->isRoomFull() && this->status == GameStatus::CONNECTING;
}


std::queue<int> Room::getPlayerSockets() const {
    std::queue<int> q;

    for(auto it = this->players.cbegin(); it != this->players.cend(); ++it) {
        if(it->second->isOnline())
            q.push(it->second->getID());
    }

    return q;
}

bool Room::isEmpty() {
    return this->players.empty();
}

const std::queue<int> &Room::getProgress() const {
    return this->progress;
}

void Room::setProgress(std::queue<int> progress) {
    this->progress = progress;
}

bool Room::hasProgress() {
    return this->progress.size() > 0;
}

void Room::startTurn() {
    this->turn++;
    this->updateActivePlayer();
}

int Room::getTurn() const {
    return this->turn;
}

void Room::deregisterPlayer(Player *p) {
    int uid = p->getID();

    int i = 0;
    for (auto const& o : this->playerOrder) {
        int id = this->playerOrder[i];
        if(id == uid) break;
        i++;
    }

    // i contains index in playerOrder
    this->playerOrder.erase(this->playerOrder.begin() + i);
    this->players.erase(p->getID());

}

int Room::countOnlinePlayers() const {
    int onlinePlayers = 0;

    for(auto it = this->players.cbegin(); it != this->players.cend(); ++it) {
        if(it->second->isOnline())
            onlinePlayers++;
    }

    return onlinePlayers;
}

bool Room::isAnyoneOnline() const {
    return this->countOnlinePlayers() > 0;
}

int Room::getWinnerID() {
    return this->winnerID;
}

void Room::finishGame() {
    int index = (this->turn + 1) % this->getPlayerCount();
    this->winnerID = this->players[this->playerOrder[index]]->getID();
    this->changeStatus(GameStatus::FINISHED);
}

bool Room::hasGameFinished() {
    return this->winnerID != -1;
}

void Room::endGame() {
    //this->changeStatus(GameStatus::FINISHED);
}

void Room::changeStatus(GameStatus s) {
    this->status = s;

    // -- log --
    std::string str = ""; str.append("Status of the room (");
    str.append(std::to_string(this->getID())); str.append(") has changed to ");
    str.append(translateGameStatus(s)); str.append("."); Logger::info(str);
}

bool Room::isReady() {
    return this->status == GameStatus::READY;
}

bool Room::checkReadyToContinue(bool replay) {
    if(this->status == GameStatus::ENDED)
        return false;

    if(this->type == GameType::SINGLEPLAYER) {

        if(replay) {
            this->changeStatus(GameStatus::FINISHED_REPLAY);
        } else {
            this->changeStatus(GameStatus::FINISHED_END);
        }

        return replay;
    }

    // multiplayer:
    if(this->status == GameStatus::FINISHED) {

        if(replay) {
            this->changeStatus(GameStatus::FINISHED_REPLAY);
        } else {
            this->changeStatus(GameStatus::FINISHED_END);
        }

        // only one checked
        return false;
    }

    if(this->status == GameStatus::FINISHED_REPLAY) {
        return true;
    }

    // both checked
    this->changeStatus(GameStatus::ENDED);
    return true;
}

bool Room::isReplayReady() {
    return this->status == GameStatus::FINISHED_REPLAY;
}

bool Room::hasGameEnded() {
    return this->status == GameStatus::ENDED;
}

int Room::getTime() const {
    /*
     *  difficulty (0 - 2): 1 - 2
     *  dimension  (1 - 5): 1 - 2
     */

    double  turn     = this->getTurn(),
            turnCoef = (turn > 5) ? turn * 0.8 : (turn > 10) ? turn * 0.5 : (turn > 15) ? turn * 0.6 : turn,
            diff     = (double) this->getDifficulty(),
            diffCoef = (diff == 2) ? 2 : 1,
            dim      = (double) this->getBoardDimension(),
            dimCoef  = (dim > 3) ? 2 : 1;
    int     time     = (int) (turnCoef * (diffCoef + dimCoef));

    //time += 500;

    time += 50;

    return this->turn == 1 ? time + Game::FIRST_TURN_RESERVE : time;
}

void Room::reassignPlayer(Player *player, Player *prevPlayer) {
    // room status?
    // players
    // player order

    this->changeStatus(GameStatus::READY);

    // this->players: PlayerMap[uid] = player
    this->deregisterPlayer(prevPlayer);
    this->registerPlayer(player);
}

GameStatus Room::getStatus() {
    return this->status;
}

bool Room::hasDifficulty() {
    return this->difficulty >= GameDifficulty::EASY && this->difficulty <= GameDifficulty::EXPERT;
}

bool Room::hasDimension() {
    return this->boardDimension >= BoardDimension::BOARD_TINY && this->boardDimension <= BoardDimension::BOARD_HUGE;
}

bool Room::hasType() {
    return this->type >= GameType::SINGLEPLAYER && this->type <= GameType::MULTIPLAYER;
}

File: aoa_client_mvn/src/main/java/cz/kiv/ups/model/Room.java
package cz.kiv.ups.model;

import cz.kiv.ups.game.BoardDimension;
import cz.kiv.ups.game.GameDifficulty;
import cz.kiv.ups.game.GameType;

public class Room {
	
	private int ID;

	private int playerCount;

	private int turn;

	private GameType type;

	private GameDifficulty difficulty;

	private BoardDimension boardDimension;

	private Player[] players;

	private  Player currentPlayer;

	private  Player opponent;



	public Room() {
	    this(-1);
	}

	public Room(int ID) {
		this.ID = ID;
	}


	public Room(GameType type, GameDifficulty diff, BoardDimension dim) {
		this();
		this.type = type;
		this.difficulty = diff;
		this.boardDimension = dim;
	}

    /**
	 * @return the id
	 */
	public int getID() {
		return ID;
	}

	/**
	 * @param id the ID to set
	 */
	public void setID(int id) {
		ID = id;
	}

	/**
	 * @return the playerCount
	 */
	public int getPlayerCount() {
		return playerCount;
	}

	/**
	 * @param playerCount the playerCount to set
	 */
	public void setPlayerCount(int playerCount) {
		this.playerCount = playerCount;
	}

	/**
	 * @return the difficulty
	 */
	public GameDifficulty getDifficulty() {
		return difficulty;
	}

	/**
	 * @param difficulty the difficulty to set
	 */
	public void setDifficulty(GameDifficulty difficulty) {
		this.difficulty = difficulty;
	}

	/**
	 * @return the boardDimension
	 */
	public BoardDimension getBoardDimension() {
		return boardDimension;
	}

	/**
	 * @param boardDimension the boardDimension to set
	 */
	public void setBoardDimension(BoardDimension boardDimension) {
		this.boardDimension = boardDimension;
	}

	/**
	 * @return the players
	 */
	public Player[] getPlayers() {
		return players;
	}

	/**
	 * @param players the players to set
	 */
	public void setPlayers(Player[] players) {
		this.players = players;
	}

	public GameType getType() {
		return type;
	}

	public void setType(GameType type) {
		this.type = type;
	}

    public int getTurn() {
        return turn;
    }

    public void setTurn(int turn) {
        this.turn = turn;
    }

	public void updatePlayers(Player[] players) {
        Player opponent = null;

		for (Player p : players) {
			if(p.getID() != this.currentPlayer.getID())
				opponent = p;
		}

		if(opponent == null)
		    return;

		this.players = players;


		this.updateOpponent();


		// determine: player has gone offline
        this.opponent.setOnline(opponent.isOnline());

        // determine: player has left a room
        this.setPlayerCount(players.length);

	}

	public boolean isOpponentOnline(){
	    return this.getType() == GameType.SINGLEPLAYER || this.opponent.isOnline();
	    //return this.opponent.isOnline();
    }

    public void setCurrentPlayer(Player currentPlayer) {
        this.currentPlayer = currentPlayer;

        for (int i = 0; i < this.players.length; i++) {
			if(this.players[i].getName().equals(this.currentPlayer.getName())) {
				this.players[i] = this.currentPlayer;
				break;
			}
		}

        this.updateOpponent();
    }

    private void updateOpponent(){
		for (Player p : this.players) {
			if(p.getID() != this.currentPlayer.getID())
				this.opponent = p;
		}
	}

    public int getCurrentPlayerID() {
        return this.currentPlayer.getID();
    }
}
